### PR TITLE
fixes #22055 - update bookmark loading status.

### PR DESCRIPTION
**Pull request by ohadlevy**
This PR both fixes the issue reported by not showing spinner if
there is an empty set of saved bookmarks (it shows None found)

It also adds missing tests cases for spinner, error and empty result sets, and replaces the Loader component with the patternfly-react Spinner component instead.

### CODE REVIEW

**Review by sharvit — APPROVED** [review]
Looks great!

File: webpack/assets/javascripts/react_app/components/bookmarks/index.js [hunk 1]
```diff
@@ -45,12 +41,17 @@ class BookmarkContainer extends React.Component {
...
+            </MenuItem>
           )}
-          {getLoadingStatus(errors, bookmarks) === STATUS.RESOLVED &&
-            bookmarks.map(({ name, query }) => <Bookmark key={name} text={name} query={query} />)}
-          {getLoadingStatus(errors, bookmarks) === STATUS.ERROR &&
+          {status === STATUS.RESOLVED &&
+            ((bookmarks.length > 0 &&
+              bookmarks.map(({ name, query }) => (
+                <Bookmark key={name} text={name} query={query} />
+              ))) || <MenuItem> __('None found')</MenuItem>)}
```
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/bookmarks/index.js]: This needs to be interpolated, i.e. `<MenuItem>{__('None found')}</MenuItem>`.

image

In addition, I think this text should not be clickable. Maybe it's worth marking it as [disabled](https://getbootstrap.com/docs/3.3/components/#dropdowns-disabled).

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ohadlevy: Thanks. I found one small issue, otherwise: 👍

**Comment by ohadlevy** [comment]
Good find! I've updated both (both disabled and extracted strings).

**Comment by ohadlevy** [comment]
@timogoebel green again

File: webpack/assets/javascripts/react_app/components/bookmarks/index.js [hunk 2]
```diff
@@ -45,12 +41,17 @@ class BookmarkContainer extends React.Component {
         <Dropdown.Toggle />
         <Dropdown.Menu>
           <MenuItem header>{__('Saved Bookmarks')}</MenuItem>
-          {getLoadingStatus(errors, bookmarks) === STATUS.PENDING && (
-            <Loader spinnerSize="xs" status={STATUS.PENDING} />
+          {status === STATUS.PENDING && (
+            <MenuItem className='loader-root'>
```
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/bookmarks/index.js]: I think we should add the spinner as a direct child element to the `<li>` in the DOM and drop the `<a>` around it.

File: webpack/assets/javascripts/react_app/redux/reducers/bookmarks/index.js [hunk 3]
```diff
@@ -43,7 +46,9 @@ export default (state = initialState, action) => {
     case BOOKMARKS_MODAL_CLOSED:
       return state.set('showModal', false);
     case BOOKMARKS_FAILURE:
-      return state.setIn([payload.controller, 'errors'], payload.error);
+      return state
+        .setIn([payload.controller, 'errors'], payload.error)
```
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/redux/reducers/bookmarks/index.js]: This does not work as well, as `payload.controller` is not defined.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ohadlevy: Thanks. Looks good. I just have one remark regarding the spinner. I also found, that the error action is not working correctly. Could you take another look, please? Thanks.

**Comment by ohadlevy** [comment]
@timogoebel fixed both issues - thanks!

**Review by timogoebel — APPROVED** [review]
👍 if tests are 🍏 .

**Comment by ohadlevy** [comment]
It's green :)

**Comment by timogoebel** [comment]
Merged, thanks @ohadlevy.

**Pull request merged**